File: minesweeper/java-minesweeper/src/test/java/it/chicio/minesweeper/field/parser/FieldValidRowParserTest.java
package it.chicio.minesweeper.field.parser;

import it.chicio.minesweeper.field.Field;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentMatchers;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class FieldValidRowParserTest {
    private FieldRowContentParser fieldRowContentParser;
    private FieldValidRowParser fieldValidRowParser;

    @Before
    public void setUp() {
        fieldRowContentParser = mock(FieldRowContentParser.class);
        fieldValidRowParser = new FieldValidRowParser(fieldRowContentParser);
    }

    @Test
    public void parseAValidRow() {
        String newRow = ". *";
        when(fieldRowContentParser.tryToParseRowAndUpdate(ArgumentMatchers.any(FieldsParsingStatus.class)))
                .thenReturn(new FieldsParsingStatusBuilder()
                        .withCurrentRowContent(newRow)
                        .withCurrentField(new Field(new String[][]{{"*", "."}, {".", "*"}}))
                        .build()
                );

        FieldsParsingStatus newFieldsParsingStatus = fieldValidRowParser.parse(
            newRow,
            new FieldsParsingStatusBuilder().build()
        );

        assertThat(newFieldsParsingStatus.currentRowContent, is(equalTo(newRow)));
        assertThat(newFieldsParsingStatus.currentField, is(equalTo(new Field(new String[][]{{"*", "."}, {".", "*"}}))));
    }

    @Test
    public void parseInvalidRow() {
        String currentRowContent = "* *";
        FieldsParsingStatus newFieldsParsingStatus = fieldValidRowParser.parse(
            "",
            new FieldsParsingStatusBuilder().withCurrentRowContent(currentRowContent).build()
        );

        assertThat(newFieldsParsingStatus.currentRowContent, is(equalTo(currentRowContent)));
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/test/kotlin/it/chicio/minesweeper/field/parser/FieldValidRowParserTest.kt
package it.chicio.minesweeper.field.parser

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import it.chicio.minesweeper.FieldFactory
import it.chicio.minesweeper.FieldsParsingStatusBuilder
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@DisplayName("FieldValidRowParser")
@ExtendWith(MockKExtension::class)
class FieldValidRowParserTest {
    @MockK
    private lateinit var fieldRowContentParser: FieldRowContentParser

    private lateinit var fieldValidRowParser: FieldValidRowParser

    @BeforeEach
    fun setUp() {
        fieldValidRowParser = FieldValidRowParser(fieldRowContentParser)
    }

    @Nested
    @DisplayName("parse")
    inner class Parse {
        @Test
        fun `a valid row`() {
            every { fieldRowContentParser.tryToParseRowAndUpdate(any()) } returns fieldsParsingStatus

            val newFieldsParsingStatus = fieldValidRowParser.parse(". *", FieldsParsingStatusBuilder().build())

            assertEquals(newFieldsParsingStatus.currentRowContent, ". *")
            assertEquals(newFieldsParsingStatus.currentField, FieldFactory().make(
                    arrayOf(
                            arrayOf("*", "."),
                            arrayOf(".", "*")
                    )
            ))
            verify(exactly = 1) { fieldRowContentParser.tryToParseRowAndUpdate(any()) }
        }

        @Test
        fun `an invalid row`() =
                assertEquals(
                        fieldValidRowParser.parse("", FieldsParsingStatusBuilder(currentRowContent = "* *").build()).currentRowContent,
                        "* *"
                )
    }

    companion object {
        private val fieldsParsingStatus = FieldsParsingStatusBuilder(
                currentRowContent = ". *",
                currentField = FieldFactory().make(
                        arrayOf(
                                arrayOf("*", "."),
                                arrayOf(".", "*")
                        )
                )
        ).build()
    }
}